naga/src/valid/analyzer.rs

```diff
@@ -1158,6 +1209,54 @@ impl FunctionInfo {
         }
         Ok(combined_uniformity)
     }
+
+    fn try_update_mesh_vertex_type(
+        &mut self,
+        ty: Handle<crate::Type>,
+        value: Handle<crate::Expression>,
+    ) -> Result<(), WithSpan<FunctionError>> {
+        if let &Some(ref existing) = &self.mesh_shader_info.vertex_type {
+            if existing.0 != ty {
+                return Err(
+                    FunctionError::ConflictingMeshOutputTypes(existing.1, value).with_span()
+                );
+            }
+        } else {
+            self.mesh_shader_info.vertex_type = Some((ty, value));
+        }
+        Ok(())
+    }
+
+    fn try_update_mesh_primitive_type(
+        &mut self,
+        ty: Handle<crate::Type>,
+        value: Handle<crate::Expression>,
+    ) -> Result<(), WithSpan<FunctionError>> {
+        if let &Some(ref existing) = &self.mesh_shader_info.primitive_type {
+            if existing.0 != ty {
+                return Err(
+                    FunctionError::ConflictingMeshOutputTypes(existing.1, value).with_span()
+                );
+            }
+        } else {
+            self.mesh_shader_info.primitive_type = Some((ty, value));
+        }
+        Ok(())
+    }
+
+    /// Update this function's mesh shader info, given that it calls `callee`.
+    fn try_update_mesh_info(
+        &mut self,
+        callee: &FunctionMeshShaderInfo,
+    ) -> Result<(), WithSpan<FunctionError>> {
+        if let &Some(ref other_vertex) = &callee.vertex_type {
+            self.try_update_mesh_vertex_type(other_vertex.0, other_vertex.1)?;
+        }
+        if let &Some(ref other_primitive) = &callee.vertex_type {
```

Comment:
Shouldn't this be:

```suggestion
        if let &Some(ref other_primitive) = &callee.primitive_type {
```

Let's add a test to `naga/tests/naga/validation.rs` for this, too.

Reply:
I'm not sure what such a test would look like